### PR TITLE
[Feature](function) Add function strcmp

### LINKED ISSUE #33093: [Enhancement](doris-future) Support strcmp function
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

mysql support [strcmp](https://dev.mysql.com/doc/refman/8.3/en/string-comparison-functions.html#function_strcmp) function. we need to support it too.

### Solution

To support a function, we need:
1. function implementation and registration in BE
2. function signature and visitor for nereids planner in FE
3. function signature in doris_builtin_functions.py

you can reference like https://github.com/apache/doris/pull/33005/files

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by xiedeyantu**
## Proposed changes

Issue Number: close #33093

```
mysql> select strcmp("test1", "test");
+-------------------------+
| strcmp('test1', 'test') |
+-------------------------+
|                       1 |
+-------------------------+
1 row in set (0.02 sec)

mysql> select strcmp("test", "test");
+------------------------+
| strcmp('test', 'test') |
+------------------------+
|                      0 |
+------------------------+
1 row in set (0.00 sec)

mysql> select strcmp("test", "test1");
+-------------------------+
| strcmp('test', 'test1') |
+-------------------------+
|                      -1 |
+-------------------------+
1 row in set (0.00 sec)
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xiedeyantu** [comment]
run buildall

**Comment by xiedeyantu** [comment]
run p0

**Comment by xiedeyantu** [comment]
run p0

**Comment by dataroaring** [comment]
run buildall

**Comment by xiedeyantu** [comment]
run p0

**Comment by xiedeyantu** [comment]
run buildall

**Comment by xiedeyantu** [comment]
@zclllyybb please review this pr.

**Comment by zclllyybb** [comment]
please add my wechat torch-wood_ for talking about this pr

File: be/src/vec/functions/function_string.h [hunk 1]
```diff
@@ -294,6 +294,62 @@ struct SubstringUtil {
...
+                assert_cast<const ColumnString&>(*block.get_by_position(arguments[1]).column);
+
+        auto result_column = ColumnInt16::create(input_rows_count);
+        auto& result_data = result_column->get_data();
+        auto null_column = ColumnUInt8::create(input_rows_count);
+        auto& null_map = null_column->get_data();
+
+        for (int row = 0; row < input_rows_count; row++) {
+            null_map[row] = false;
+            if (arg0_column.is_nullable() || arg1_column.is_nullable()) {
```
Comment (zclllyybb) [inline be/src/vec/functions/function_string.h:325]: no this

File: be/src/vec/functions/function_string.h [hunk 2]
```diff
@@ -294,6 +294,62 @@ struct SubstringUtil {
...
+    static constexpr auto name = "strcmp";
+
+    static FunctionPtr create() { return std::make_shared<FunctionStrcmp>(); }
+
+    String get_name() const override { return name; }
+
+    size_t get_number_of_arguments() const override { return 2; }
+
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        return std::make_shared<DataTypeInt16>();
```
Comment (zclllyybb) [inline be/src/vec/functions/function_string.h:308]: check nullable property and wrap it.

File: be/src/vec/functions/function_string.h [hunk 3]
```diff
@@ -294,6 +294,62 @@ struct SubstringUtil {
...
+    }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) const override {
+        const auto& arg0_column =
```
Comment (zclllyybb) [inline be/src/vec/functions/function_string.h:313]: deal the const logic

File: regression-test/suites/query_p0/sql_functions/string_functions/test_string_function.groovy
Comment (zclllyybb) [inline regression-test/suites/query_p0/sql_functions/string_functions/test_string_function.groovy:1]: delete the regression-test cases but add them to be-ut

**Review by zclllyybb — CHANGES_REQUESTED** [review]

**Comment by xiedeyantu** [comment]
run buildall

**Comment by xiedeyantu** [comment]
run beut

**Comment by xiedeyantu** [comment]
run buildall

**Comment by xiedeyantu** [comment]
run buildall

File: be/src/vec/functions/function_string.h [hunk 4]
```diff
@@ -294,6 +294,46 @@ struct SubstringUtil {
...
+        const auto& [arg0_column, arg0_const] =
+                unpack_if_const(block.get_by_position(arguments[0]).column);
+        const auto& [arg1_column, arg1_const] =
+                unpack_if_const(block.get_by_position(arguments[1]).column);
+
+        auto result_column = ColumnInt16::create(input_rows_count);
+        auto& result_data = result_column->get_data();
+
+        for (int row = 0; row < input_rows_count; row++) {
+            auto arg0 = arg0_column->get_data_at(row);
```
Comment (zclllyybb) [inline be/src/vec/functions/function_string.h:325]: implement `vector_vector` and `vector_const`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ScalarFunctionVisitor.java [hunk 5]
```diff
@@ -2011,6 +2012,10 @@ default R visitInttoUuid(InttoUuid inttoUuid, C context) {
         return visitScalarFunction(inttoUuid, context);
     }
 
+    default R visitStrcmp(Strcmp strcmp, C context) {
+        return visitScalarFunction(strcmp, context);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ScalarFunctionVisitor.java:2016]: add to legacy planner too~

File: fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java [hunk 6]
```diff
@@ -628,6 +629,7 @@ public class BuiltinScalarFunctions implements FunctionHelper {
             scalar(InnerProduct.class, "inner_product"),
             scalar(Instr.class, "instr"),
             scalar(InttoUuid.class, "int_to_uuid"),
+            scalar(Strcmp.class, "strcmp"),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java:632]: add with lexicographical order

Comment (xiedeyantu) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java:632]: > add with lexicographical order

ok

**Comment by xiedeyantu** [comment]
run buildall

**Comment by xiedeyantu** [comment]
run buildall

File: be/test/vec/function/function_string_test.cpp [hunk 7]
```diff
@@ -1187,4 +1187,24 @@ TEST(function_string_test, function_uuid_test) {
     }
 }
 
+TEST(function_string_test, function_strcmp_test) {
+    std::string func_name = "strcmp";
+    InputTypeSet input_types = {TypeIndex::String, TypeIndex::String};
+
+    DataSet data_set = {{{Null(), Null()}, Null()},
```
Comment (zclllyybb) [inline be/test/vec/function/function_string_test.cpp:1194]: add some test for const columns

**Comment by xiedeyantu** [comment]
run buildall

**Comment by xiedeyantu** [comment]
run buildall

**Comment by xiedeyantu** [comment]
run buildall

**Review by zclllyybb — APPROVED** [review]
LGTM

**Review by morrySnow — DISMISSED** [review]

File: be/src/vec/functions/function_string.h [hunk 8]
```diff
@@ -294,6 +294,72 @@ struct SubstringUtil {
...
+    static constexpr auto name = "strcmp";
+
+    static FunctionPtr create() { return std::make_shared<FunctionStrcmp>(); }
+
+    String get_name() const override { return name; }
+
+    size_t get_number_of_arguments() const override { return 2; }
+
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        if (arguments[0]->is_nullable() || arguments[1]->is_nullable()) {
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.h:308]: no need consider nulllable in `use_default_null`

File: be/src/vec/functions/function_string.h [hunk 9]
```diff
@@ -294,6 +294,72 @@ struct SubstringUtil {
...
+            return make_nullable(std::make_shared<DataTypeInt16>());
+        }
+        return std::make_shared<DataTypeInt16>();
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.h:311]: why not use Tinyint ?

**Review by HappenLee — CHANGES_REQUESTED** [review]
check the return type

File: be/src/vec/functions/function_string.h (hunk 8 above)
Comment (xiedeyantu) [reply be/src/vec/functions/function_string.h]: fixed

File: be/src/vec/functions/function_string.h (hunk 9 above)
Comment (xiedeyantu) [reply be/src/vec/functions/function_string.h]: @HappenLee fixed, please review

**Comment by xiedeyantu** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by zhangstar333 — APPROVED** [review]

**Comment by zhangstar333** [comment]
run cloud_p1

**Pull request merged**